parquet-hadoop/src/main/java/org/apache/parquet/filter2/compat/PredicateEvaluation.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.parquet.filter2.compat;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.apache.parquet.filter2.predicate.FilterPredicate;
+import org.apache.parquet.filter2.predicate.Operators;
+
+/**
+ * Used in Filters to mark whether we should DROP the block if data matches the condition.
+ * If we cannot decide whether the block matches, it will be always safe to return BLOCK_MIGHT_MATCH.
+ * We use Boolean Object here to distinguish the value type, please do not modify it.
+ */
+public class PredicateEvaluation {
+  /* The block might match, but we cannot decide yet, will check in the other filters. */
+  public static final Boolean BLOCK_MIGHT_MATCH = new Boolean(false);
+  /* The block can match for sure. */
+  public static final Boolean BLOCK_MUST_MATCH = new Boolean(false);
+  /* The block can't match for sure */
+  public static final Boolean BLOCK_CANNOT_MATCH = new Boolean(true);
+
+  public static Boolean evaluateAnd(Operators.And and, FilterPredicate.Visitor<Boolean> predicate) {
+    Boolean left = and.getLeft().accept(predicate);
+    if (left == BLOCK_CANNOT_MATCH) {
+      // seems unintuitive to put an || not an && here but we can
+      // drop a chunk of records if we know that either the left or
+      // the right predicate agrees that no matter what we don't
+      // need this chunk.
+      return BLOCK_CANNOT_MATCH;
+    }
+    Boolean right = and.getRight().accept(predicate);
+    if (right == BLOCK_CANNOT_MATCH) {
+      return BLOCK_CANNOT_MATCH;
+    } else if (left == BLOCK_MUST_MATCH && right == BLOCK_MUST_MATCH) {
```

Comment:
```suggestion
    } else if (left == BLOCK_MUST_MATCH || right == BLOCK_MUST_MATCH) {
```

Reply:
Both left and right are not `BLOCK_CANNOT_MATCH`, so it can return `BLOCK_MUST_MATCH ` if either side is `BLOCK_MUST_MATCH`

Reply:
if left is `BLOCK_MUST_MATCH` , right is might `BLOCK_MIGHT_MATCH` , left & right should be `BLOCK_MIGHT_MATCH`.
Because in the next filter may right may be `BLOCK_CANNOT_MATCH ` and we should drop it

Reply:
I didn't get your point here. @yabola 

If the current expression is `A and B`, then following result applies regardless of other expressions:
- A is BLOCK_MUST_MATCH and B is BLOCK_MUST_MATCH => BLOCK_MUST_MATCH
- A is BLOCK_MUST_MATCH and B is BLOCK_MIGHT_MATCH => BLOCK_MUST_MATCH
- A is BLOCK_MIGHT_MATCH and B is BLOCK_MUST_MATCH => BLOCK_MUST_MATCH
- A is BLOCK_MIGHT_MATCH and B is BLOCK_MIGHT_MATCH => BLOCK_MIGHT_MATCH
- A is BLOCK_CANNOT_MATCH or/and B is BLOCK_CANNOT_MATCH => BLOCK_CANNOT_MATCH